<a href="https://colab.research.google.com/github/ElofssonLab/kb8029-book/blob/main/notebooks/day07-discussion-2.ipynb" style="display:inline-block;padding:10px 18px;background-color:#F9AB00;color:#000000;font-weight:bold;text-decoration:none;border-radius:6px;font-family:sans-serif;font-size:14px;">&#9654;&nbsp; Open in Google Colab</a>

# Day 7 — In-class discussion problem (2 of 3)

Work this out **by hand in your group first** — then run the code cell
to check your answer before presenting.

## What happens when two training examples disagree?

A perceptron with one input, no bias ($z = wx$), squared-error loss,
starting at $w_0 = 0$, learning rate $\eta = 0.02$ — but now **two**
training examples at once, $(x,y) = (2,6)$ and $(x,y) = (3,3)$, with
total loss $L(w) = (2w-6)^2 + (3w-3)^2$.

**As a group:**

1. Is there any single value of $w$ that makes *both* predictions exact
   ($2w=6$ and $3w=3$ simultaneously)?
2. Compute $dL/dw$ at $w=0$ by hand, and take one gradient descent step.
3. Predict: where does $w$ eventually converge, if there's no value that
   fits both examples perfectly?

Run the cell below to check your answers.

In [1]:
def grad(w):
    # d/dw [(2w-6)^2 + (3w-3)^2]
    return 2*(2*w - 6)*2 + 2*(3*w - 3)*3

w = 0.0
eta = 0.02
g = grad(w)
w1 = w - eta * g
print(f"gradient at w=0: {g}")
print(f"after one step: w = {w1}")

# where gradient descent eventually settles: dL/dw = 0
w_star = 42 / 26
print(f"\noptimal w* (minimizes total squared error): {w_star:.4f}")
print(f"prediction for x=2 at w*: {2*w_star:.3f}  (target was 6)")
print(f"prediction for x=3 at w*: {3*w_star:.3f}  (target was 3)")

gradient at w=0: -42.0
after one step: w = 0.84

optimal w* (minimizes total squared error): 1.6154
prediction for x=2 at w*: 3.231  (target was 6)
prediction for x=3 at w*: 4.846  (target was 3)


**Discussion point:** no single weight satisfies both examples —
$2w=6$ needs $w=3$, but $3w=3$ needs $w=1$, and one weight can't be both.
Gradient descent doesn't get stuck or fail here; it converges to
$w^* \approx 1.615$, the value that minimizes the *total* squared error
across both examples — a genuine compromise, matching neither target
exactly but doing the best it can with the capacity it has (one
weight, no bias). This is exactly the situation real datasets are in
all the time: not every example can be fit perfectly, and the model
settles on the best available balance. It's also a preview of why more
capacity (more weights, hidden layers) sometimes matters — not because
the model was "broken," but because one weight is a very small amount
of capacity to work with.